src/test/java/com/iexec/common/utils/ContextualLockRunnerTests.java

```diff
@@ -0,0 +1,377 @@
+
+        boolean hasRun = locks.getWithLock(1, () -> true);
+
+        Assertions.assertThat(hasRun).isTrue();
+    }
+
+    @Test
+    void applyWithLock() {
+        final ContextualLockRunner<Integer> locks = new ContextualLockRunner<>();
+
+        Wrapper<Integer> wrappedValue = new Wrapper<>(0);
+        boolean hasRun = locks.applyWithLock(1, wrappedValue::setValueAndAcknowledge);
+
+        Assertions.assertThat(hasRun).isTrue();
+        Assertions.assertThat(wrappedValue.getValue()).isEqualTo(1);
+    }
+
+    @Data
+    @AllArgsConstructor
+    private static class Wrapper<T> {
+        private T value;
+
+        public boolean setValueAndAcknowledge(T value) {
+            this.value = value;
+            return true;
+        }
+    }
+    // endregion
+
+    // region Check synchronization works
+    /**
+     * A constant value is used to lock all operations:
+     * final order should be a list within which each value should not interfere with others.
+     * E.g. [2, 2, 2, 1, 1, 1, 0, 0, 0, 3, 3, 3] would be an expected result,
+     * however [2, <b>1</b>, 2, 2, 1, 1, 0, 0, 0, 3, 3, 3] would not be an expected result as 1 interferes with 2.
+     */
+    @Test
+    void runWithLockOnConstantValue() {
+        runWithLock(i -> true);
+        Assertions.assertThatThrownBy(() -> runWithoutLock(i -> true))
+                .getRootCause()
+                .hasMessageContaining("Synchronization failed");
+    }
+
+    /**
+     * Parity is used to lock all operations:
+     * final order should be a list within which each value should not interfere with others of same parity.
+     * E.g. [2, 1, 2, 2, 0, 1, 1, 3, 0, 0, 0, 3, 3] would be an expected result,
+     * however [2, 1, 2, 2, 1, <b>3</b>, 1, 0, 0, 0, 3, 3] would not be an expected result as 3 interferes with 1.
+     */
+    @Test
+    void runWithLockOnParity() {
+        runWithLock(i -> i % 2 == 0);
+        Assertions.assertThatThrownBy(() -> runWithoutLock(i -> i % 2 == 0))
+                .getRootCause()
+                .hasMessageContaining("Synchronization failed");
+    }
+
+    /**
+     * Operation's value are used to lock operations:
+     * any order is then acceptable as we don't care about interferences between distinct operations.
+     */
+    @Test
+    void runWithLockPerValue() {
+        runWithLock(Function.identity());
+
+        // As any order is acceptable, there should not be any exception there.
+        Assertions.assertThatCode(() -> runWithoutLock(Function.identity()))
+                .doesNotThrowAnyException();
+    }
+
+    /**
+     * Parity is used to lock all operations:
+     * final order should be a list within which each value should not interfere with others of same parity.
+     * E.g. [2, 1, 2, 2, 0, 1, 1, 3, 0, 0, 0, 3, 3] would be an expected result,
+     * however [2, 1, 2, 2, 1, <b>3</b>, 1, 0, 0, 0, 3, 3] would not be an expected result as 3 interferes with 1.
```

Comment:
Are those comments still valid after tests refactoring ?
New algorithm does not seem to check execution order, it seems to check concurrent decrement on a key in a map.
Do not hesitate to correct me if I am wrong

Reply:
They are not valid anymore, true. Let's find a new way to describe what's done there.

Reply:
See https://github.com/iExecBlockchainComputing/iexec-common/pull/295/commits/2c408a0b41f068509502f1eaa50e6257a792359d.